### PR TITLE
Refactor VisitLambda to not create a child walker

### LINKED ISSUE #44411: CS8602 CS8604 analysis are not correct after Visual Studio is upgrated to 16.6.0
**Version Used**: 

Visual Studio 16.6.0

**Steps to Reproduce**:

```csharp
class Program
{
    static void Main(string[] args)
    {
        var foo = GetNotNullObject();

        Task.Run(() => Bar());
        Task.Run(() => Baz());

        void Bar() => VerifyNotNull(foo);
        void Baz() => Console.WriteLine(foo.ToString());
    }

    private static object GetNotNullObject() => new object();

    private static bool VerifyNotNull(object instance) => instance != null;
}
```

Just create such code in Visual Studio 16.6.0 and you'll get the wrong CS8602 and CS8604 warning. Before I upgrade to 16.6.0 (e.g. 16.5.4 ) the analysis works fine.

See the minimal reproduction repo: [walterlv/Walterlv.Issues.CSharpNullableCS8604](https://github.com/walterlv/Walterlv.Issues.CSharpNullableCS8604)

**Expected Behavior**:

The `foo` variable is analyzed as not null.

**Actual Behavior**:

The `foo` variable is analyzed as nullable and there is a warning during the compiling.

image

### LINKED ISSUE #45795: DoesNotReturn analysis complains about 'return' statements within local functions
**Version Used**: 7b4bfaaee

**Steps to Reproduce**:

```cs
using System.Diagnostics.CodeAnalysis;
#nullable enable
public class C {
    [DoesNotReturn]
    public void M() {
        _ = local1();
        throw null!;
        
        int local1() { return 1; }
    }
}
```

**Expected Behavior**: No warnings

**Actual Behavior**: Warning on 'return 1' within local function

### PR DESCRIPTION
**Pull request by RikkiGibson**
Fixes #44411
Fixes #45795

I had hoped to outright remove the `_useDelegateInvokeParameterTypes` field and ctor parameter, as well as the `delegateInvokeMethodOpt` ctor parameter, but I cannot prove to my satisfaction that they will never make a difference in the one scenario where we create a NullableWalker specifically from a lambda, which is `UnboundLambda.GetInferredReturnType`. There are no tests which fail when the `delegateInvokeMethodOpt` is forced to `null` in this context.

I also found that after this change, we were throwing on ImmutableDictionary creation in the bootstrap build, because dictionary values of type `TypeWithAnnotations` were being compared for equality. This is solved by manually passing a value equality comparer when creating the dictionary.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -50494,6 +50499,64 @@ static void G(object? x)
...
+    }
+
+    private static object GetNotNullObject() => new object();
+
+    private static bool VerifyNotNull(object instance) => instance != null;
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics();
+        }
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:50558]: Extra testing I think we need here:

- Tests which verify that null state changes in a lambda do not persist after the lambda is evaluated
- Tests which verify that null state changes in local functions do not persist after the local function is evaluated
- Tests which verify the null state of local functions is the union of all the call sites. Essentially have a variable which is `maybe-null` in one call and `non-null` in the other and verify with in that it's `maybe-null`.  This should be tested inside a lambda as well as outside. Possible the latter test already exists.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:50558]: It is likely that these tests already exist, I can try to identify them and follow up if this is the case.

Comment (jaredpar) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:50558]: I agree it's likely they do but want to ensure we have them given that this is the crux of the problem we've been discussing. If we can identify them I think we should annotate them with a reference to this GitHub issue so that it's easy for future developers to see the relationship. If not then ... well there isn't too much harm in having the same test twice with a different name :smile:

Comment (jaredpar) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:50558]: We should also make sure that we have tests for anonymous methods ... cause someone is going to use them with nullable.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -2145,31 +2145,56 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
...
+            Debug.Assert(!useDelegateInvokeParameterTypes || delegateInvokeMethod is object);
+            var oldDelegateInvokeMethod = _delegateInvokeMethod;
+            _delegateInvokeMethod = delegateInvokeMethod;
+            var oldUseDelegateInvokeParameterTypes = _useDelegateInvokeParameterTypes;
+            _useDelegateInvokeParameterTypes = useDelegateInvokeParameterTypes;
+
+            var savedState = this.State;
+            this.State = state;
+
+            var oldPending = SavePending(); // we do not support branches into a lambda
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:2195]: Feel like this comment says "what", can u make it say "why"?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:50558]: I tried identifying tests for these scenarios by simply breaking the relevant part of the NullableWalker and reading the failed test output. For all 3 I didn't feel like I found a test which succinctly demonstrated each scenario, so I am adding the tests.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -2179,24 +2204,24 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
...
 
-            if (!localFuncSymbol.Locations.IsDefaultOrEmpty)
+            if (!lambdaOrFunctionSymbol.Locations.IsDefaultOrEmpty)
             {
-                location = localFuncSymbol.Locations[0];
+                location = lambdaOrFunctionSymbol.Locations[0];
             }
 
-            LeaveParameters(localFuncSymbol.Parameters, localFunc.Syntax, location);
+            LeaveParameters(lambdaOrFunctionSymbol.Parameters, lambdaOrFunction.Syntax, location);
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:2224]: Good question! I found that the IsCompilerGenerated check went by unremarked when it was introduced: https://github.com/dotnet/roslyn/pull/40422/files#diff-e3717bc6a837e218d79277c0e45f213aR1620

One thing I did notice is that LeaveParameters does absolutely nothing in NullableWalker, it only has an effect in DefiniteAssignment.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 4]
```diff
@@ -50494,6 +50499,226 @@ static void G(object? x)
...
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (16,9): warning CS8602: Dereference of a possibly null reference.
+                //         x.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(16, 9));
+        }
+
+        [Fact, WorkItem(44411, "https://github.com/dotnet/roslyn/issues/44411")]
+        public void LocalFunctionWritesDoNotUpdateContainingMethodState()
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:50657]: The behavior of this test gave me pause, particularly due to this comment in the original code:

```cs
// State changes to captured variables are recorded, as calls to local functions
// transition the state of captured variables if the variables have state changes
// across all branches leaving the local function
```

However, the behavior of this test does not differ in master, therefore I am inclined to file a separate bug for it.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: I feel the doc comment on SavePending is adequate in explaining why we call it here, therefore will delete the comment.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 4 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:50657]: Which part of the behavior is giving u pause here?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:50657]: The comment claims that calls to local functions will transition the state of captured variables if the variables have state changes across all branches leaving the local function. I interpreted this to mean that e.g. if you unconditionally assign 'null' to a captured variable as in this test, it's expected that the state of the captured variable will change to maybe-null at call sites to the local function.

Comment (jaredpar) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:50657]: Agree. We should file a follow up issue.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 5]
```diff
@@ -133,30 +135,28 @@ public VisitArgumentResult(VisitResult visitResult, Optional<LocalState> stateFo
...
-        /// Use the parameter types and nullability from _methodSignatureOpt for initial
-        /// parameter state. If false, the signature of _member is used instead.
+        /// If true, the parameter types and nullability from _delegateInvokeMethod is used for
+        /// initial parameter state. If false, the signature of CurrentSymbol is used instead.
         /// </summary>
-        private readonly bool _useDelegateInvokeParameterTypes;
+        private bool _useDelegateInvokeParameterTypes;
 
         /// <summary>
-        /// Method signature used for return type or parameter types. Distinct from _member
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:142]: >or parameter types [](start = 50, length = 18)

It looks like `_delegateInvokeMethod` is also used for parameters.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 6]
```diff
@@ -1889,7 +1889,7 @@ protected override LocalState ReachableBottomState()
 
         private void EnterParameters()
         {
-            if (!(_symbol is MethodSymbol methodSymbol))
+            if (!(CurrentSymbol is MethodSymbol methodSymbol))
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:1892]: >CurrentSymbol [](start = 18, length = 13)

There are other references to `_symbol` that are not updated. Are those other references correct?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 7]
```diff
@@ -2145,31 +2145,56 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
...
                 if (Symbol.IsCaptured(symbol, localFunc.Symbol))
                 {
-                    State[slot] = localFunctionState.StartingState[slot];
+                    state[slot] = localFunctionState.StartingState[slot];
                 }
             }
             localFunctionState.Visited = true;
 
-            if (!localFunc.WasCompilerGenerated) EnterParameters(localFuncSymbol.Parameters);
+            AnalyzeLocalFunctionOrLambda(
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:2164]: >AnalyzeLocalFunctionOrLambda [](start = 12, length = 28)

Are we saving and restoring `_returnTypesOpt`?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 8]
```diff
@@ -2145,31 +2145,56 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
...
             var localFuncSymbol = localFunc.Symbol;
-            var delegateInvokeMethod = _delegateInvokeMethod;
-            this.CurrentSymbol = localFuncSymbol;
-            _delegateInvokeMethod = null;
-
-            var oldPending = SavePending(); // we do not support branches into a lambda
-
-            var savedState = this.State;
-            var localFunctionState = GetOrCreateLocalFuncUsages(localFuncSymbol);
+            var localFunctionState = GetOrCreateLocalFuncUsages(localFunc.Symbol);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:2149]: >localFunc.Symbol [](start = 64, length = 16)

Is this change needed?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 9]
```diff
@@ -2145,31 +2145,56 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
 
         public override BoundNode? VisitLocalFunctionStatement(BoundLocalFunctionStatement localFunc)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:2146]: >localFunc [](start = 91, length = 9)

Not part of this change, but consider naming parameter `node`, to match other `Visit` methods and to remove the confusion between `localFunc` and `localFuncSymbol`.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 7 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:2164]: Hmm, it does seem like maybe this save-restore should be pushed into `AnalyzeLocalFunctionOrLambda`. I'll try and write a test which fails with this current implementation.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 10]
```diff
@@ -2179,24 +2204,24 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
...
             RestorePending(oldPending);
 
             Location? location = null;
 
-            if (!localFuncSymbol.Locations.IsDefaultOrEmpty)
+            if (!lambdaOrFunctionSymbol.Locations.IsDefaultOrEmpty)
             {
-                location = localFuncSymbol.Locations[0];
+                location = lambdaOrFunctionSymbol.Locations[0];
             }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:2222]: >} [](start = 12, length = 1)

`var location = lambdaOrFunctionSymbol.Locations.FirstOrNone();`

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 11]
```diff
@@ -5196,7 +5218,7 @@ private VariableState GetVariableState(Optional<LocalState> localState)
             return new VariableState(
                 _variableSlot.ToImmutableDictionary(),
                 ImmutableArray.Create(variableBySlot, start: 0, length: nextVariableSlot),
-                _variableTypes.ToImmutableDictionary(),
+                _variableTypes.ToImmutableDictionary(_variableTypes.Comparer, TypeWithAnnotations.EqualsComparer.ConsiderEverythingComparer),
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5221]: >_variableTypes.Comparer, TypeWithAnnotations.EqualsComparer.ConsiderEverythingComparer [](start = 53, length = 86)

Is this change needed? (The key comparer is the default, and the value comparer doesn't appear to be used.)

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5221]: Yes, without this change, the bootstrap build crashes because it attempts to call `TypeWithAnnotations.Equals` on the dictionary values.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 12]
```diff
@@ -50494,6 +50499,226 @@ static void G(object? x)
...
+        {
+            var source = @"
+using System.Threading.Tasks;
+using System;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        var foo = GetNotNullObject();
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:50513]: >foo [](start = 12, length = 3)

Consider using a different name, here and test below.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 11 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5221]: Where are we calling `Equals` on the dictionary values?

---
In reply to: [451162769](https://github.com/dotnet/roslyn/pull/45527#discussion_r451162769) [](ancestors = 451162769)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 13]
```diff
@@ -50494,6 +50499,226 @@ static void G(object? x)
...
+        x.ToString();
+
+        void local()
+        {
+            x = null;
+        }
+    }
+
+    static void M2(Action a) { }
+}
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:50682]: Please add a similar test where the local function is defined before `x` is used in the containing method.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 11 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5221]: `ToImmutableDictionary()` may call Equals on dictionary values internally when constructing the immutable dictionary. This is something we might want to examine in greater detail. It's not obvious to me why sharing the `_variableTypes` between a lambda and its containing method would make this behavior start to surface in the bootstrap build.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 14]
```diff
@@ -50494,6 +50499,226 @@ static void G(object? x)
...
+        }
+    }
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (15,17): warning CS8602: Dereference of a possibly null reference.
+                //             _ = x.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(15, 17));
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:50720]: >} [](start = 8, length = 1)

Please add a similar test where the local function is passed as a delegate.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 15]
```diff
@@ -2145,31 +2145,56 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
...
+            var oldSymbol = this.CurrentSymbol;
+            this.CurrentSymbol = lambdaOrFunctionSymbol;
+
+            Debug.Assert(!useDelegateInvokeParameterTypes || delegateInvokeMethod is object);
+            var oldDelegateInvokeMethod = _delegateInvokeMethod;
+            _delegateInvokeMethod = delegateInvokeMethod;
+            var oldUseDelegateInvokeParameterTypes = _useDelegateInvokeParameterTypes;
+            _useDelegateInvokeParameterTypes = useDelegateInvokeParameterTypes;
+
+            var savedState = this.State;
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Done

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 4 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:50657]: Filed #45770

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 16]
```diff
@@ -731,7 +731,8 @@ int getSlotForFieldOrProperty(Symbol member)
 
         private void EnforceDoesNotReturn(SyntaxNode? syntaxOpt)
         {
-            if (_symbol is MethodSymbol method &&
+            // DoesNotReturn is only supported in member methods
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:734]: >DoesNotReturn [](start = 15, length = 13)

Can `[DoesNotReturn]` be applied to a local function?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 17]
```diff
@@ -30425,6 +30425,82 @@ void Main(object? o)
...
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source, DoesNotReturnAttributeDefinition });
+            comp.VerifyDiagnostics();
+        }
+
+        [Theory]
+        [InlineData("")]
+        [InlineData("static ")]
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:30452]: Does `[DoesNotReturn]` analysis depend on whether the member is `static`?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 18]
```diff
@@ -30425,6 +30425,82 @@ void Main(object? o)
...
+using System.Diagnostics.CodeAnalysis;
+class D
+{
+    void Main(object? o)
+    {
+        boom();
+        [DoesNotReturn]
+        " + modifiers + @"void boom()
+        {
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:30496]: >} [](start = 8, length = 1)

Should report `WRN_ShouldNotReturn`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 19]
```diff
@@ -30425,6 +30425,82 @@ void Main(object? o)
...
+        [InlineData("")]
+        [InlineData("static ")]
+        [WorkItem(37081, "https://github.com/dotnet/roslyn/issues/37081")]
+        public void DoesNotReturn_LocalFunction_CheckImpl(string modifiers)
+        {
+            string source = @"
+using System.Diagnostics.CodeAnalysis;
+class D
+{
+    void Main(object? o)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:30490]: >object? o [](start = 14, length = 9)

Not used.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 16 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:734]: We don't complain if it is applied, but we do not change analysis when it is present.

I actually expected us to mark regions unreachable in NullableWalker when calling a local function that does not return, and thus give no warnings in those regions--definite assignment already works such that the following scenario gives no errors. Unfortunately this does not seem to be the case. It's an inconsistency we might want to follow up on.

```cs
string s;
local1();
s.ToString();

void local1() => throw null!;
```

```cs
string? s = null;
local1();
s.ToString(); // this warns, but shouldn't

void local1() => throw null!;
```

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 6 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:1893]: I made a pass over existing usages of `_symbol` and changed the ones that seemed wrong.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 7 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:2165]: Done

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 16 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:734]: I will try and ensure that DoesNotReturn does not impact analysis of local functions for the moment, and file a follow-up issue to confirm that this is expected.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 17 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:30452]: No, I pushed this by accident, I believe this function should be deleted.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:30452]: Simplified the test down to a `[Fact]` to simplify, since 'static' differences are not expected to impact the test.

**Comment by 333fred** [comment]
>                 if (_symbol is MethodSymbol method)

Why should this not be `CurrentSymbol`?

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:476 in c5b1eec. [](commit_id = c5b1eec9e65eb7e4e706f951c5b756a9b319d40b, deletion_comment = False)

**Comment by 333fred** [comment]
>                 if (_symbol is MethodSymbol method)

Why should this not be `CurrentSymbol`?

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:524 in c5b1eec. [](commit_id = c5b1eec9e65eb7e4e706f951c5b756a9b319d40b, deletion_comment = False)

**Comment by 333fred** [comment]
>                 if (_symbol is MethodSymbol method)

Why should this not be `CurrentSymbol`?

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:568 in c5b1eec. [](commit_id = c5b1eec9e65eb7e4e706f951c5b756a9b319d40b, deletion_comment = False)

**Comment by 333fred** [comment]
>                             var method = getTopLevelMethod(_symbol as MethodSymbol);

It seems like `_symbol` here should always be the top level method? Is there a time when it's not? If there is, the docs on `_symbol` are probably wrong.

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:1218 in c5b1eec. [](commit_id = c5b1eec9e65eb7e4e706f951c5b756a9b319d40b, deletion_comment = False)

**Comment by RikkiGibson** [comment]
> 
> 
> > ```
> >                         var method = getTopLevelMethod(_symbol as MethodSymbol);
> > ```
> 
> It seems like `_symbol` here should always be the top level method? Is there a time when it's not? If there is, the docs on `_symbol` are probably wrong.
> 
> Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:1218 in [c5b1eec](https://github.com/dotnet/roslyn/commit/c5b1eec9e65eb7e4e706f951c5b756a9b319d40b). [](commit_id = [c5b1eec](https://github.com/dotnet/roslyn/commit/c5b1eec9e65eb7e4e706f951c5b756a9b319d40b), deletion_comment = False)

`_symbol` is not the top level method when we are reinferring a lambda return type.

**Comment by RikkiGibson** [comment]
> 
> 
> > ```
> >             if (_symbol is MethodSymbol method)
> > ```
> 
> Why should this not be `CurrentSymbol`?
> 
> Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:476 in [c5b1eec](https://github.com/dotnet/roslyn/commit/c5b1eec9e65eb7e4e706f951c5b756a9b319d40b). [](commit_id = [c5b1eec](https://github.com/dotnet/roslyn/commit/c5b1eec9e65eb7e4e706f951c5b756a9b319d40b), deletion_comment = False)

I would prefer to address these cases in a follow-up PR which resolves #45821

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 20]
```diff
@@ -2143,33 +2144,61 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
             }
         }
 
-        public override BoundNode? VisitLocalFunctionStatement(BoundLocalFunctionStatement localFunc)
+        public override BoundNode? VisitLocalFunctionStatement(BoundLocalFunctionStatement node)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:2147]: Alright, so I think speculative analysis in local functions was already broken in current master: we're never calling `snapshotBuilderOpt?.EnterNewWalker()` on the local function symbol, so state saving and restore will be broken.

**Comment by RikkiGibson** [comment]
It appears that tests already verify that we respect `[DoesNotReturn]` on local functions. Therefore I will revert the behavior and delete my contradictory test.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 21]
```diff
@@ -6876,50 +6897,51 @@ public override BoundNode VisitLambda(BoundLambda node)
...
-            else if (_snapshotBuilderOpt is object && delegateTypeOpt is object && !node.UnboundLambda.HasExplicitlyTypedParameterList)
+
+            var delegateInvokeMethod = delegateTypeOpt?.DelegateInvokeMethod;
+            bool useDelegateInvokeParameterTypes = UseDelegateInvokeParameterTypes(node, delegateInvokeMethod);
+            if (useDelegateInvokeParameterTypes && _snapshotBuilderOpt is object)
             {
                 SetUpdatedSymbol(node, node.Symbol, delegateTypeOpt);
             }
 
-            Analyze(
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:6902]: By no longer calling this analyze overload, we're not entering a new scope for the lambda, so speculation will be broken. We need to restore the calls to `snapshotManager?.EnterNewWalker` with the new symbol.

**Comment by 333fred** [comment]
Done review pass (commit 7). I didn't look at tests.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 20 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:2147]: This should reproduce in that variables declared in the local function don't have correct state when restoring for semantic analysis (if previous statements in the local function checked them for null, for example.

---
In reply to: [451908958](https://github.com/dotnet/roslyn/pull/45527#discussion_r451908958) [](ancestors = 451908958)

Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:2147]: Opened https://github.com/dotnet/roslyn/issues/45825 to track this. I think that in order to not introduce this bug for lambdas in this PR you're going to have to fix this one as well.
